File: btap-necb/scripts/legacy_whatsnew.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# What has moved in the legacy fork since our pinned oracle?
#
# The gems' only tie to openstudio-standards is `legacy_pin/REF`, ONE commit SHA.
# It is not a branch: bundler fetches the fork by SHA. So "are we behind?" is not
# a question git can answer from inside this repository — nothing here points at
# the fork's history. This script answers it.
#
# It reports the commits between the pin and a fork branch (default `nrcan`),
# and maps each changed path to the gem that should care, so a reviewer or an
# LLM can judge what is worth absorbing before running the bump workflow in
# legacy_pin/README.md.
#
#   rake legacy:whatsnew                     # against nrcan
#   BRANCH=develop rake legacy:whatsnew      # against another branch
#   LEGACY_FORK=/path/to/openstudio-standards rake legacy:whatsnew
#
# Resolution order for the fork, mirroring legacy_pin/Gemfile's own convention:
#   LEGACY_FORK -> LEGACY_PIN_REMOTE (when it is a local path) -> a cached
#   blobless mirror of the GitHub remote under tmp/.
#
# READ-ONLY. It never changes REF, never bundles, never merges.

require 'fileutils'
require 'shellwords'

ROOT = File.expand_path('../..', __dir__)
PIN_DIR = File.join(ROOT, 'legacy_pin')
REF = File.read(File.join(PIN_DIR, 'REF')).strip
BRANCH = ENV['BRANCH'] || 'nrcan'
DEFAULT_REMOTE = 'https://github.com/NatLabRockies/openstudio-standards.git'
CACHE = File.join(ROOT, 'tmp', 'legacy_fork.git')

# Which gem owns a legacy path. Heuristic and deliberately coarse — it points a
# reviewer at the right gem, it does not decide anything. First match wins, so
# the costing sheets are claimed before the per-domain patterns.
# LOCATION rules come first: a file under test/ is a legacy test artifact even
# when its NAME contains 'shw-scale' or 'chiller-type'. Letting the keyword
# patterns match first filed regression fixtures under the domain gems, which
# reads as "this gem has work to do" when it does not.
OWNERS = [
  [%r{^test/},                                                     'legacy TESTS — oracle behaviour, not our source'],
  [%r{^data/},                                                     'legacy DATA — check whether a gem vendored a copy'],
  [%r{/btap/costing/|/btap/common/.*\.csv$}, 'costing (hvac + envelope + lighting + shw)'],
  [%r{/btap/geometry\.rb|create_shape|create_bar},                'btap-modeling'],
  [%r{space_types\.json|schedules\.json|beps_compliance_path\.rb}, 'btap-necb (loads)'],
  [%r{lighting|daylight},                                          'btap-necb (lighting)'],
  [%r{service_water_heating|shw},                                  'btap-necb (shw)'],
  [%r{hvac_system|autozone|efficienc|curves|chiller|boiler|/fan},  'btap-necb (hvac) or btap-modeling'],
  [%r{building_envelope|thermal_transmittance|fdwr|thermal_bridging}, 'btap-necb (envelope)'],
  [%r{necb_20\d\d\.rb},                                            'btap-necb (+ the domain each hunk touches)']
].freeze

def die(msg)
  warn msg
  exit 1
end

def git(dir, *args)
  out = IO.popen([{ 'GIT_TERMINAL_PROMPT' => '0' }, 'git', '-C', dir, *args], &:read)
  [$?.success?, out]
end

# A local checkout beats any network path: it is instant and works offline.
def local_fork
  %w[LEGACY_FORK LEGACY_PIN_REMOTE].each do |var|
    path = ENV[var]
    next if path.nil? || path.empty? || path.start_with?('http', 'git@')
    next unless Dir.exist?(File.join(path, '.git')) || Dir.exist?(File.join(path, 'objects'))

    return File.expand_path(path)
  end
  nil
end

# Blobless mirror: commits and trees only, so --name-status works without
# dragging the fork's blob history across the wire. GitHub honours the filter;
# a plain file:// or dumb server does NOT (git prints "filtering not recognized
# by server, ignoring" and clones everything, ~4.5 GB for this fork). Prefer a
# local checkout via LEGACY_FORK when you have one — it needs no copy at all.
def cached_mirror
  remote = ENV['LEGACY_FORK'] || DEFAULT_REMOTE
  unless Dir.exist?(CACHE)
    FileUtils.mkdir_p(File.dirname(CACHE))
    warn "fetching a blobless mirror of #{remote} into #{CACHE} (first run only)..."
    ok = system('git', 'clone', '--bare', '--filter=blob:none', remote, CACHE)
    die("could not clone #{remote} — pass LEGACY_FORK=/path/to/a/local/checkout instead") unless ok
  end
  system('git', '-C', CACHE, 'fetch', '--quiet', 'origin', "+refs/heads/#{BRANCH}:refs/heads/#{BRANCH}")
  CACHE
end

fork_dir = local_fork
source = fork_dir ? "local checkout #{fork_dir}" : 'cached mirror of the GitHub remote'
fork_dir ||= cached_mirror

ok, = git(fork_dir, 'cat-file', '-e', "#{REF}^{commit}")
die("the pinned revision #{REF[0, 12]} is not in #{fork_dir} — wrong repository?") unless ok

# Prefer the REMOTE-TRACKING ref. In a working checkout the local branch is
# routinely stale (it is only as fresh as the last `git checkout` + pull), and
# "what has moved upstream" is a question about the remote. Fall back to the
# local branch when there is no remote ref — e.g. in the bare mirror.
resolve = ->(cand) { git(fork_dir, 'rev-parse', '--verify', '--quiet', "#{cand}^{commit}") }
remote_ok, remote_sha = resolve.call("origin/#{BRANCH}")
local_ok,  local_sha  = resolve.call("refs/heads/#{BRANCH}")
tip = remote_ok ? "origin/#{BRANCH}" : (local_ok ? "refs/heads/#{BRANCH}" : nil)
die("branch '#{BRANCH}' not found in #{fork_dir} (try BRANCH=develop)") unless tip

if remote_ok && local_ok && remote_sha.strip != local_sha.strip
  warn "note: #{fork_dir} has a local '#{BRANCH}' at #{local_sha.strip[0, 9]} that differs from " \
       "origin/#{BRANCH} at #{remote_sha.strip[0, 9]} — reporting against the remote " \
       '(run `git fetch` in the fork for the freshest answer)'
end

_, pin_line = git(fork_dir, 'log', '-1', '--format=%h %ad %s', '--date=short', REF)
_, tip_line = git(fork_dir, 'log', '-1', '--format=%h %ad %s', '--date=short', tip)
_, behind = git(fork_dir, 'rev-list', '--count', "#{REF}..#{tip}")
_, ahead  = git(fork_dir, 'rev-list', '--count', "#{tip}..#{REF}")

puts '=' * 78
puts "legacy oracle: pinned at #{pin_line.strip}"
puts "fork branch:   #{BRANCH} at #{tip_line.strip}"
puts "source:        #{source}"
puts '=' * 78

if behind.to_i.zero?
  puts "\nThe pin is CURRENT with #{BRANCH}." \
       "#{ahead.to_i.positive? ? " (#{ahead.strip} commit(s) on the pin are not on #{BRANCH})" : ''}"
  exit 0
end

puts "\n#{behind.strip} commit(s) on #{BRANCH} since the pin:\n\n"
_, log = git(fork_dir, 'log', '--format=  %h  %ad  %s', '--date=short', '--reverse', "#{REF}..#{tip}")
puts log

_, names = git(fork_dir, 'diff', '--name-status', REF, tip)
changed = names.lines.map { |l| l.chomp.split("\t", 2) }.reject { |p| p.size < 2 }

grouped = Hash.new { |h, k| h[k] = [] }
changed.each do |status, path|
  owner = OWNERS.find { |re, _| path =~ re }&.last || 'UNMAPPED — review by hand'
  grouped[owner] << "#{status}  #{path}"
end

puts "\n#{changed.size} path(s) changed, by the gem that should care:\n"
# Legacy tests and data last: they are context, not work.
order = ->(k) { k.start_with?('legacy ') ? 1 : (k.start_with?('UNMAPPED') ? 2 : 0) }
grouped.sort_by { |k, v| [order.call(k), -v.size] }.each do |owner, paths|
  puts "\n  #{owner}  (#{paths.size})"
  paths.first(12).each { |p| puts "      #{p}" }
  puts "      ... #{paths.size - 12} more" if paths.size > 12
end

puts <<~NEXT

  #{'-' * 78}
  Absorbing any of this means BUMPING THE PIN, not copying code:
    1. edit legacy_pin/REF to the new fork revision
    2. BUNDLE_GEMFILE=legacy_pin/Gemfile bundle install
    3. run every parity gate with LEGACY_PIN_REQUIRED=1, plus the archetype
       regeneration gates
    4. a flipped assertion means the upstream change reached a COMPARED
       behaviour — attribute each one before accepting it (the D-68 discipline)
  See legacy_pin/README.md.
NEXT

File: python/scripts/legacy_whatsnew.py
#!/usr/bin/env python3
"""Report commits and changed paths between legacy_pin/REF and a fork tip."""

from __future__ import annotations

import argparse
import json
import os
import re
import subprocess
import sys
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parents[2]
PIN_REF = REPO_ROOT / "legacy_pin" / "REF"
DEFAULT_REMOTE = "https://github.com/NatLabRockies/openstudio-standards.git"
DEFAULT_CACHE = REPO_ROOT / "tmp" / "legacy_fork.git"

OWNERS = [
    (re.compile(r"^test/"), "legacy TESTS - oracle behaviour, not our source"),
    (re.compile(r"^data/"), "legacy DATA - check whether Python vendored a copy"),
    (re.compile(r"/btap/costing/|/btap/common/.*\.csv$"),
     "costing (hvac + envelope + lighting + shw)"),
    (re.compile(r"/btap/geometry\.rb|create_shape|create_bar"), "btap.modeling"),
    (re.compile(r"space_types\.json|schedules\.json|beps_compliance_path\.rb"),
     "btap.necb (loads)"),
    (re.compile(r"lighting|daylight"), "btap.necb (lighting)"),
    (re.compile(r"service_water_heating|shw"), "btap.necb (shw)"),
    (re.compile(r"hvac_system|autozone|efficienc|curves|chiller|boiler|/fan"),
     "btap.necb (hvac) or btap.modeling"),
    (re.compile(r"building_envelope|thermal_transmittance|fdwr|thermal_bridging"),
     "btap.necb (envelope)"),
    (re.compile(r"necb_20\d\d\.rb"), "btap.necb (+ the domain each hunk touches)"),
]


class LegacyError(RuntimeError):
    pass


def git(directory: Path, *args: str, check: bool = True) -> str:
    result = subprocess.run(
        ["git", "-C", str(directory), *args], capture_output=True, text=True,
        env={**os.environ, "GIT_TERMINAL_PROMPT": "0"}, check=False,
    )
    if check and result.returncode != 0:
        raise LegacyError(result.stderr.strip() or f"git {' '.join(args)} failed")
    return result.stdout


def owner_for(path: str) -> str:
    return next((owner for pattern, owner in OWNERS if pattern.search(path)),
                "UNMAPPED - review by hand")


def local_fork() -> Path | None:
    for variable in ("LEGACY_FORK", "LEGACY_PIN_REMOTE"):
        value = os.environ.get(variable, "")
        if not value or value.startswith(("http", "git@")):
            continue
        path = Path(value).expanduser().resolve()
        if (path / ".git").is_dir() or (path / "objects").is_dir():
            return path
    return None


def cached_mirror(branch: str, cache: Path = DEFAULT_CACHE) -> Path:
    remote = os.environ.get("LEGACY_FORK", DEFAULT_REMOTE)
    if not cache.is_dir():
        cache.parent.mkdir(parents=True, exist_ok=True)
        result = subprocess.run(
            ["git", "clone", "--bare", "--filter=blob:none", remote, str(cache)],
            check=False,
        )
        if result.returncode != 0:
            raise LegacyError(
                f"could not clone {remote} - pass LEGACY_FORK=/path/to/a/local/checkout")
    subprocess.run(
        ["git", "-C", str(cache), "fetch", "--quiet", "origin",
         f"+refs/heads/{branch}:refs/heads/{branch}"], check=True,
    )
    return cache


def _resolves(directory: Path, revision: str) -> bool:
    result = subprocess.run(
        ["git", "-C", str(directory), "rev-parse", "--verify", "--quiet",
         f"{revision}^{{commit}}"], capture_output=True, check=False,
    )
    return result.returncode == 0


def report(directory: Path, pin: str, *, branch: str = "nrcan",
           tip: str | None = None) -> dict:
    if not _resolves(directory, pin):
        raise LegacyError(f"pinned revision {pin[:12]} is not in {directory}")
    if tip is None:
        tip = next((candidate for candidate in (f"origin/{branch}", f"refs/heads/{branch}")
                    if _resolves(directory, candidate)), None)
    if tip is None or not _resolves(directory, tip):
        raise LegacyError(f"branch or tip '{tip or branch}' not found in {directory}")
    tip_sha = git(directory, "rev-parse", f"{tip}^{{commit}}").strip()
    commits = [line.split("\t", 3) for line in git(
        directory, "log", "--format=%H%x09%h%x09%ad%x09%s", "--date=short",
        "--reverse", f"{pin}..{tip_sha}").splitlines()]
    changed = []
    for line in git(directory, "diff", "--name-status", pin, tip_sha).splitlines():
        parts = line.split("\t")
        if len(parts) < 2:
            continue
        path = parts[-1]
        changed.append({"status": parts[0], "path": path, "owner": owner_for(path)})
    return {"pin": pin, "tip": tip_sha, "branch": branch,
            "commits": [{"sha": row[0], "short": row[1], "date": row[2],
                         "subject": row[3]} for row in commits],
            "changed": changed}


def main(argv=None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--json", action="store_true")
    parser.add_argument("--branch", default=os.environ.get("BRANCH", "nrcan"))
    parser.add_argument("--tip", default=os.environ.get("TIP"))
    args = parser.parse_args(argv)
    directory = local_fork() or cached_mirror(args.branch)
    result = report(directory, PIN_REF.read_text(encoding="utf-8").strip(),
                    branch=args.branch, tip=args.tip)
    if args.json:
        print(json.dumps(result, indent=1))
        return 0
    print(f"legacy oracle: {result['pin']}")
    print(f"fork tip:     {result['tip']} ({result['branch']})")
    print(f"{len(result['commits'])} commit(s), {len(result['changed'])} changed path(s)")
    for commit in result["commits"]:
        print(f"  {commit['short']}  {commit['date']}  {commit['subject']}")
    grouped: dict[str, list[dict]] = {}
    for item in result["changed"]:
        grouped.setdefault(item["owner"], []).append(item)
    for owner, items in sorted(grouped.items(), key=lambda pair: pair[0]):
        print(f"\n{owner} ({len(items)})")
        for item in items:
            print(f"  {item['status']}  {item['path']}")
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except LegacyError as error:
        print(error, file=sys.stderr)
        raise SystemExit(1) from error